### PR TITLE
Add UT for ReconcileVPCSubnet and its subfunctions

### PR DESCRIPTION
**Pull request by Shilpa-Gokul**
Add UT for ReconcileVPCSubnet, SetVPCSubnetStatus, checkVPCSubnet, createVPCSubnet functions
Partially fixes https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/1818

### CODE REVIEW

File: cloud/scope/powervs_cluster_test.go [hunk 1]
```diff
@@ -1704,3 +1704,512 @@ func TestCreateServiceInstance(t *testing.T) {
...
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When GetVPCSubnetByName returns empty as vpc subnet doesnt exist", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1993]:
```suggestion
	t.Run("When GetVPCSubnetByName returns empty as vpc subnet doesnot exist", func(t *testing.T) {
```

**Review by Karthik-K-N — COMMENTED** [review]
few suggestions, Otherwise LGTM

File: cloud/scope/powervs_cluster_test.go [hunk 2]
```diff
@@ -1704,3 +1704,498 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		subnet1Options := &vpcv1.GetSubnetOptions{ID: subnet1Details.ID}
+		mockVPC.EXPECT().GetSubnet(subnet1Options).Return(subnet1Details, nil, nil)
+		subnet2Details := &vpcv1.Subnet{ID: ptr.To("subnet2ID"), Name: ptr.To("subnet2Name")}
+		subnet2Options := &vpcv1.GetSubnetOptions{ID: subnet2Details.ID}
+		mockVPC.EXPECT().GetSubnet(subnet2Options).Return(subnet2Details, nil, nil)
+		instance, err := clusterScope.ReconcileVPCSubnets()
+		g.Expect(instance).To(BeFalse())
+		g.Expect(err).To(BeNil())
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.VPCSubnet[*subnet1Details.Name].ID).To(Equal(subnet1Details.ID))
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.VPCSubnet[*subnet2Details.Name].ID).To(Equal(subnet2Details.ID))
```
[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:1744]: I could see that we are not setting controllercreated [here](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_cluster.go#L1150) so i thought its not needed

[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go:1744]: we are not setting but its expected to be false, so better to add a check

[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:1744]: Done, added check for nil

File: cloud/scope/powervs_cluster_test.go [hunk 3]
```diff
@@ -1704,3 +1704,512 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		subnet1Options := &vpcv1.GetSubnetOptions{ID: subnet1Details.ID}
+		mockVPC.EXPECT().GetSubnet(subnet1Options).Return(subnet1Details, nil, nil)
+		subnet2Details := &vpcv1.Subnet{ID: ptr.To("subnet2ID"), Name: ptr.To("subnet2Name")}
+		subnet2Options := &vpcv1.GetSubnetOptions{ID: subnet2Details.ID}
+		mockVPC.EXPECT().GetSubnet(subnet2Options).Return(subnet2Details, nil, nil)
+		instance, err := clusterScope.ReconcileVPCSubnets()
+		g.Expect(instance).To(BeFalse())
+		g.Expect(err).To(BeNil())
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.VPCSubnet["subnet1Name"].ID).To(Equal(subnet1Details.ID))
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.VPCSubnet["subnet2Name"].ID).To(Equal(subnet2Details.ID))
```
[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Addressed

File: cloud/scope/powervs_cluster_test.go [hunk 4]
```diff
@@ -1704,3 +1704,512 @@ func TestCreateServiceInstance(t *testing.T) {
...
+	})
+
+	t.Run("When VPCSubnets are not set in spec and vpc doesnt exist already, create new", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Done, changed as suggested

**Comment by Shilpa-Gokul** [comment]
@Karthik-K-N @Prajyot-Parab @Amulyam24 Have addressed all the comments, could you please review again?

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

/assign @Amulyam24 @Prajyot-Parab

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm
/hold for @Amulyam24

**Comment by Shilpa-Gokul** [comment]
> @Shilpa-Gokul Is this dependent on any of the other PRs you are working on or ready to merge? /cc @Amulyam24

@Prajyot-Parab This is not dependent on any other PRs and is ready to merge.

File: cloud/scope/powervs_cluster_test.go [hunk 5]
```diff
@@ -1704,3 +1704,499 @@ func TestCreateServiceInstance(t *testing.T) {
...
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When GetVPCSubnetByName returns empty as vpc subnet doesnot exist", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1981]:
```suggestion
	t.Run("When GetVPCSubnetByName returns nil as vpc subnet does not exist in cloud", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 6]
```diff
@@ -1704,3 +1704,499 @@ func TestCreateServiceInstance(t *testing.T) {
...
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")},
+					VPC: &infrav1beta2.VPCResourceReference{Region: ptr.To("aadd")}},
+			},
+		}
+		subnet := infrav1beta2.Subnet{Name: ptr.To("ClusterName-vpcsubnet-eu-de-1")}
+		subnetID, err := clusterScope.createVPCSubnet(subnet)
+		g.Expect(subnetID).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+	t.Run("When VPCZonesForVPCRegion returns no vpczones", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:2185]:
```suggestion
	t.Run("When VPCZonesForVPCRegion returns zero zones for the region set in spec", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 7]
```diff
@@ -1704,3 +1704,499 @@ func TestCreateServiceInstance(t *testing.T) {
...
+					VPCSubnets: []infrav1beta2.Subnet{{ID: ptr.To("subnet1ID"), Name: ptr.To("subnet1Name")}, {ID: ptr.To("subnet2ID"), Name: ptr.To("subnet2Name")}}},
+			},
+		}
+		subnet1Details := &vpcv1.Subnet{ID: ptr.To("subnet1ID"), Name: ptr.To("subnet1Name")}
+		subnet1Options := &vpcv1.GetSubnetOptions{ID: subnet1Details.ID}
+		mockVPC.EXPECT().GetSubnet(subnet1Options).Return(subnet1Details, nil, nil)
+		subnet2Details := &vpcv1.Subnet{ID: ptr.To("subnet2ID"), Name: ptr.To("subnet2Name")}
+		subnet2Options := &vpcv1.GetSubnetOptions{ID: subnet2Details.ID}
+		mockVPC.EXPECT().GetSubnet(subnet2Options).Return(subnet2Details, nil, nil)
+		instance, err := clusterScope.ReconcileVPCSubnets()
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1740]:
```suggestion
		requeue, err := clusterScope.ReconcileVPCSubnets()
```
Please change wherever applicable

**Review by Amulyam24 — COMMENTED** [review]
Minor suggestions, LGTM!

**Comment by Shilpa-Gokul** [comment]
@Prajyot-Parab @Amulyam24 @Karthik-K-N Have addressed all the comments, please take a look

**Review by Amulyam24 — APPROVED** [review]
/lgtm

/unhold

**Review by Amulyam24 — COMMENTED** [review]
/lgtm

**Pull request merged**